### PR TITLE
[AArch64] Introduce new armv9.6 features

### PR DESCRIPTION
**Pull request by Lukacma**
This patch implements new features introduced in 2024 release of ARM ISA and creates predicates, which will be used by new instructions. 

Co-authored-by: Caroline Concatto caroline.concatto@arm.com
Co-authored-by: Spencer Abson spencer.abson@arm.com

### CODE REVIEW

File: llvm/lib/Target/AArch64/AArch64Features.td [hunk 1]
```diff
@@ -522,6 +522,39 @@ def FeatureTLBIW : ExtensionWithMArch<"tlbiw", "TLBIW", "FEAT_TLBIW",
...
+
+def FeatureF8F16MM: ExtensionWithMArch<"f8f16mm", "F8F16MM", "FEAT_F8F16MM",
+  "Enable FP8 to Half-Precision Matrix Multiplication", [FeatureFP8DOT2, FeatureF8F32MM]>;
+
+def FeatureFPRCVT: ExtensionWithMArch<"fprcvt", "FPRCVT", "FEAT_FPRCVT",
+  "Enable A64 base convert instructions for SIMD&FP scalar register operands of"
+  " different input and output sizes", [FeatureFPARMv8]>;
+
+def FeatureLSFE : ExtensionWithMArch<"lsfe", "LSFE", "FEAT_LSFE",
+  "Enable A64 base Atomic floating-point in-memory instructions", [FeatureFPARMv8]>;
```
Comment (jthackray) [inline llvm/lib/Target/AArch64/AArch64Features.td:539]: Can we mention "Large System Float Extension" so it's obvious what the extension stands for?

Perhaps "Enable Armv9.6-A Large System Float Extension (Atomic floating-point in-memory instructions)" ?

File: llvm/lib/Target/AArch64/AArch64Features.td [hunk 2]
```diff
@@ -522,6 +522,39 @@ def FeatureTLBIW : ExtensionWithMArch<"tlbiw", "TLBIW", "FEAT_TLBIW",
 //  Armv9.6 Architecture Extensions
 //===----------------------------------------------------------------------===//
 
+def FeatureCMPBR : ExtensionWithMArch<"cmpbr", "CMPBR", "FEAT_CMPBR",
+  "Enable A64 base compare and branch instructions">;
```
Comment (jthackray) [inline llvm/lib/Target/AArch64/AArch64Features.td:526]: Can we change to "Enable Armv9.6-A ..."?

File: llvm/lib/Target/AArch64/AArch64Features.td [hunk 3]
```diff
@@ -522,6 +522,39 @@ def FeatureTLBIW : ExtensionWithMArch<"tlbiw", "TLBIW", "FEAT_TLBIW",
...
+
+def FeatureF8F32MM: ExtensionWithMArch<"f8f32mm", "F8F32MM", "FEAT_F8F32MM",
+  "Enable FP8 to Single-Precision Matrix Multiplication", [FeatureFP8DOT4]>;
```
Comment (jthackray) [inline llvm/lib/Target/AArch64/AArch64Features.td:529]: Ditto, can we change to "Enable Armv9.6-A ..."

File: llvm/lib/Target/AArch64/AArch64Features.td [hunk 4]
```diff
@@ -522,6 +522,39 @@ def FeatureTLBIW : ExtensionWithMArch<"tlbiw", "TLBIW", "FEAT_TLBIW",
...
+
+def FeatureF8F16MM: ExtensionWithMArch<"f8f16mm", "F8F16MM", "FEAT_F8F16MM",
+  "Enable FP8 to Half-Precision Matrix Multiplication", [FeatureFP8DOT2, FeatureF8F32MM]>;
```
Comment (jthackray) [inline llvm/lib/Target/AArch64/AArch64Features.td:532]: Ditto, can we change to "Enable Armv9.6-A ..."

File: llvm/lib/Target/AArch64/AArch64Features.td [hunk 5]
```diff
@@ -522,6 +522,39 @@ def FeatureTLBIW : ExtensionWithMArch<"tlbiw", "TLBIW", "FEAT_TLBIW",
...
+
+def FeatureFPRCVT: ExtensionWithMArch<"fprcvt", "FPRCVT", "FEAT_FPRCVT",
+  "Enable A64 base convert instructions for SIMD&FP scalar register operands of"
```
Comment (jthackray) [inline llvm/lib/Target/AArch64/AArch64Features.td:535]: Ditto, can we change to "Enable Armv9.6-A ..."

File: llvm/lib/Target/AArch64/AArch64Features.td (hunk 3 above)
Comment (CarolineConcatto) [reply llvm/lib/Target/AArch64/AArch64Features.td:529]: I am not sure this is correct @jthackray. I believe FEAT_F8F32MM is not mandatory in armv9.6

File: llvm/lib/Target/AArch64/AArch64Features.td (hunk 4 above)
Comment (CarolineConcatto) [reply llvm/lib/Target/AArch64/AArch64Features.td:532]: I am not sure this is correct @jthackray. I believe FEAT_F8F16MM is not mandatory in armv9.6

File: llvm/lib/Target/AArch64/AArch64Features.td [hunk 6]
```diff
@@ -833,8 +866,8 @@ def HasV9_5aOps : Architecture64<9, 5, "a", "v9.5a",
   [HasV9_4aOps, FeatureCPA],
   !listconcat(HasV9_4aOps.DefaultExts, [FeatureCPA,  FeatureLUT, FeatureFAMINMAX])>;
 def HasV9_6aOps : Architecture64<9, 6, "a", "v9.6a",
-  [HasV9_5aOps],
-  !listconcat(HasV9_5aOps.DefaultExts, [])>;
+  [HasV9_5aOps, FeatureCMPBR, FeatureFPRCVT, FeatureSVE2p2],
+  !listconcat(HasV9_5aOps.DefaultExts, [FeatureCMPBR, FeatureFPRCVT, FeatureSVE2p2])>;
```
[the remark this answers is not in the document]
Comment (SpencerAbson) [reply llvm/lib/Target/AArch64/AArch64Features.td:870]: (Actually this is a logical consequence of our default Armv9 configuration having SVE2).

**Comment by Lukacma** [comment]
@jthackray @CarolineConcatto  Regarding using Armv9.6-A in feature descriptions, I am bit on the fence here. On the hand, I think it would make it clearer, in what architecture extension these features has been introduced, but on other hand these features can also be used with Armv9.5-A architecture and that might lead to user confusion as they might think these features are only available from Armv9.6-A  onwards. In the past, some feature used architecture extension names in their description and some didn't so not sure if there is any precedens here. In any case it might good to be consistent here.

File: llvm/lib/Target/AArch64/AArch64Features.td (hunk 1 above)
Comment (Lukacma) [reply llvm/lib/Target/AArch64/AArch64Features.td]: Done

File: llvm/lib/Target/AArch64/AArch64Features.td (hunk 3 above)
Comment (Lukacma) [reply llvm/lib/Target/AArch64/AArch64Features.td]: Done

File: llvm/unittests/TargetParser/TargetParserTest.cpp [hunk 7]
```diff
@@ -1413,6 +1424,9 @@ TEST(TargetParserTest, AArch64ExtensionFeatures) {
   EXPECT_TRUE(llvm::is_contained(Features, "+tlbiw"));
   EXPECT_TRUE(llvm::is_contained(Features, "+jsconv"));
   EXPECT_TRUE(llvm::is_contained(Features, "+complxnum"));
+  EXPECT_TRUE(llvm::is_contained(Features, "+lsfe"));
+  EXPECT_TRUE(llvm::is_contained(Features, "+fprcvt"));
+  EXPECT_TRUE(llvm::is_contained(Features, "+cmpbr"));
```
[the remark this answers is not in the document]
Comment (Lukacma) [reply llvm/unittests/TargetParser/TargetParserTest.cpp:1430]: Done

File: llvm/unittests/TargetParser/TargetParserTest.cpp [hunk 8]
```diff
@@ -1510,12 +1524,16 @@ TEST(TargetParserTest, AArch64ArchExtFeature) {
...
+      {"sve-bfscale", "nosve-bfscale", "+sve-bfscale", "-sve-bfscale"},
       {"sve2", "nosve2", "+sve2", "-sve2"},
       {"sve2-aes", "nosve2-aes", "+sve2-aes", "-sve2-aes"},
       {"sve2-sm4", "nosve2-sm4", "+sve2-sm4", "-sve2-sm4"},
       {"sve2-sha3", "nosve2-sha3", "+sve2-sha3", "-sve2-sha3"},
       {"sve2p1", "nosve2p1", "+sve2p1", "-sve2p1"},
+      {"sve2p2", "nosve2p2", "+sve2p2", "-sve2p2"},
       {"sve2-bitperm", "nosve2-bitperm", "+sve2-bitperm", "-sve2-bitperm"},
+      {"sve-aes2", "nosve-aes2", "+sve-aes2", "-sve-aes2"},
+      {"ssve-aes", "nossve-aes", "+ssve-aes", "-ssve-aes"},
```
[the remark this answers is not in the document]
Comment (Lukacma) [reply llvm/unittests/TargetParser/TargetParserTest.cpp:1538]: Done

File: llvm/unittests/TargetParser/TargetParserTest.cpp [hunk 9]
```diff
@@ -1791,22 +1822,52 @@ AArch64ExtensionDependenciesBaseArchTestParams
...
+        {AArch64::ARMV9_6A,
+         {"sve-bfscale", "nosve-b16b16"},
+         {},
+         {"sve-b16b16", "sve-bfscale"}},
+
+        // sve2p1 -> {sve2p2}
+        {AArch64::ARMV9_6A, {"nosve2p1", "sve2p2"}, {"sve2p1", "sve2p2"}, {}},
+        {AArch64::ARMV9_6A, {"sve2p2", "nosve2p1"}, {}, {"sve2p1", "sve2p2"}},
+
+        // sve2p1 -> sve-f16f32mm
```
[the remark this answers is not in the document]
Comment (Lukacma) [reply llvm/unittests/TargetParser/TargetParserTest.cpp]: Done

**Comment by Lukacma** [comment]
In the end, we are going to go forward with adding Armv9.6-A to all feature descriptions.

File: llvm/lib/Target/AArch64/AArch64Features.td [hunk 10]
```diff
@@ -522,6 +522,39 @@ def FeatureTLBIW : ExtensionWithMArch<"tlbiw", "TLBIW", "FEAT_TLBIW",
 //  Armv9.6 Architecture Extensions
 //===----------------------------------------------------------------------===//
 
+def FeatureCMPBR : ExtensionWithMArch<"cmpbr", "CMPBR", "FEAT_CMPBR",
```
Comment (CarolineConcatto) [inline llvm/lib/Target/AArch64/AArch64Features.td:525]: Hey Marian, 
If the requirements are not going to block us with the implementation of the instructions. 
I was thinking if maybe it is not better to remove all the requirements from this patch and we address this when all the dpISA is already upstream.

File: clang/test/Driver/aarch64-v96a.c [hunk 11]
```diff
@@ -6,14 +6,50 @@
...
+// RUN: %clang -target aarch64 -march=armv9.6-a+f8f32mm -### -c %s 2>&1 | FileCheck -check-prefix=V96A-F8F32MM %s
+// V96A-F8F32MM: "-cc1"{{.*}} "-triple" "aarch64{{.*}}" "-target-cpu" "generic" "-target-feature" "+v9.6a"{{.*}} "-target-feature" "+f8f32mm"
+
+// RUN: %clang -target aarch64 -march=armv9.6a+lsfe -### -c %s 2>&1 | FileCheck -check-prefix=V96A-LSFE %s
+// RUN: %clang -target aarch64 -march=armv9.6-a+lsfe -### -c %s 2>&1 | FileCheck -check-prefix=V96A-LSFE %s
+// V96A-LSFE: "-cc1"{{.*}} "-triple" "aarch64{{.*}}" "-target-cpu" "generic" "-target-feature" "+v9.6a"{{.*}} "-target-feature" "+lsfe"
+
+// RUN: %clang -target aarch64 -march=armv9.6a+sme2p2 -### -c %s 2>&1 | FileCheck -check-prefix=V96A-SME2p2 %s
+// RUN: %clang -target aarch64 -march=armv9.6-a+sme2p2 -### -c %s 2>&1 | FileCheck -check-prefix=V96A-SME2p2 %s
+// V96A-SME2p2: "-cc1"{{.*}} "-triple" "aarch64{{.*}}" "-target-cpu" "generic" "-target-feature" "+v9.6a"{{.*}} "-target-feature" "+sme2p2"
```
Comment (CarolineConcatto) [inline clang/test/Driver/aarch64-v96a.c:35]: Does this also adds as well "-target-feature" "+sme2p1"

File: clang/test/Driver/aarch64-v96a.c [hunk 12]
```diff
@@ -6,14 +6,50 @@
...
+
+// RUN: %clang -target aarch64 -march=armv9.6a+ssve-aes -### -c %s 2>&1 | FileCheck -check-prefix=V96A-SSVE-AES %s
+// RUN: %clang -target aarch64 -march=armv9.6-a+ssve-aes -### -c %s 2>&1 | FileCheck -check-prefix=V96A-SSVE-AES %s
+// V96A-SSVE-AES: "-cc1"{{.*}} "-triple" "aarch64{{.*}}" "-target-cpu" "generic" "-target-feature" "+v9.6a"{{.*}} "-target-feature" "+ssve-aes"
+
+// RUN: %clang -target aarch64 -march=armv9.6a+sve2p2 -### -c %s 2>&1 | FileCheck -check-prefix=V96A-SVE2p2 %s
+// RUN: %clang -target aarch64 -march=armv9.6-a+sve2p2 -### -c %s 2>&1 | FileCheck -check-prefix=V96A-SVE2p2 %s
+// V96A-SVE2p2: "-cc1"{{.*}} "-triple" "aarch64{{.*}}" "-target-cpu" "generic" "-target-feature" "+v9.6a"{{.*}} "-target-feature" "+sve2p2"
```
Comment (CarolineConcatto) [inline clang/test/Driver/aarch64-v96a.c:43]: Does this also adds "-target-feature" "+sme2p1"?

**Review by CarolineConcatto — APPROVED** [review]

**Pull request merged**